Extension/src/LanguageServer/extension.ts

```diff
@@ -60,6 +63,72 @@ export function activate(activationEventOccurred: boolean): void {
     }
 }
 
+export function registerCustomConfigurationProvider(provider: CustomConfigurationProvider): void {
+    customConfigurationProviders.push(provider);
+}
+
+export async function provideCustomConfiguration(document: vscode.TextDocument, client: Client): Promise<void> {
+    let tokenSource: CancellationTokenSource = new CancellationTokenSource();
+    return runBlockingThenableWithTimeout(async () => {
+        // Loop through registered providers until one is able to service the current document
+        for (let i: number = 0; i < customConfigurationProviders.length; i++) {
+            if (await customConfigurationProviders[i].canProvideConfiguration(document.uri)) {
+                return customConfigurationProviders[i].provideConfigurations([document.uri]);
+            }
+        }
+        return Promise.reject("No providers found for " + document.uri);
+    }, 1000, client, tokenSource).then((configs: SourceFileConfigurationItem[]) => {
+        if (configs !== null && configs.length > 0) {
+            client.sendCustomConfigurations(configs);
+        }
+    });
+}
+
+export function onDidCustomConfigurationChange(customConfigurationProvider: CustomConfigurationProvider): void {
+    // TODO: How to handle multiple workspaces? Only works with the current workspace.
+    // Should this be called when the workspace changes?
+    if (vscode.workspace.textDocuments === undefined || vscode.workspace.textDocuments.length === 0) {
+        return;
+    }
+
+    let documents: vscode.Uri[] = [];
+    for (let i: number = 0; i < vscode.workspace.textDocuments.length; ++i) {
```

Comment:
You're not going to want to do it this way.  This doesn't have any notion of which workspace (read: `Client`) owns which files.  Each `Client` tracks the files it owns, so you want to have each client request updates for its tracked documents and then send the updated configs to its language server.